internal/mountinfo/mountinfo.go

```diff
@@ -0,0 +1,206 @@
+		// every possible kind of storage device (e.x. logical volumes, NFS, etc.) See
+		// https://unix.stackexchange.com/a/11312 for more information.
+		//
+		//  As a result, this logic will only work correctly for mountFilePaths that are either:
+		// 	- stored directly on a block device
+		//	- stored on a block device's partition
+		//
+		// For all other device types, this logic will either:
+		// - find an incorrect device name
+		// - emit an error in the logs
+		//
+		// This logic was implemented from information gathered from the following sources (amongst others):
+		// - "The Linux Programming Interface" by Michael Kerrisk: Chapter 14
+		// - "Linux Kernel Development" by Robert Love: Chapters 13, 17
+		// - https://man7.org/linux/man-pages/man5/sysfs.5.html
+		// - https://en.wikipedia.org/wiki/Sysfs
+		// - https://unix.stackexchange.com/a/11312
+
+		// 'stat' the mount's mountFilePath path to determine what the device's ID numbers is
+		discoveryLogger := logger.Scoped("deviceNameDiscovery", "").
+			With(sglog.String("mountName", mountName)).
+			With(sglog.String("mountFilePath", mountFilePath))
+
+		discoveryLogger.Debug(
+			"'stat'-ing mountFilePath",
+			sglog.String("operation", "discovering device number"),
+		)
+
+		var stat unix.Stat_t
+		err := unix.Stat(mountFilePath, &stat)
+		if err != nil {
+			discoveryLogger.Debug("failed to stat",
+				sglog.String("operation", "discovering device number"),
+				sglog.Error(err),
+			)
+
+			continue
+		}
+
+		// extract the major and minor portions of the device ID, and
+		// represent it in <major>:<minor> format
+		major, minor := unix.Major(uint64(stat.Dev)), unix.Minor(uint64(stat.Dev))
+		deviceNumber := fmt.Sprintf("%d:%d", major, minor)
+
+		discoveryLogger.Debug("discovered device number",
+			sglog.String("operation", "discovering device number"),
+			sglog.String("deviceNumber", deviceNumber),
+		)
+
+		// /sys/dev/block/<device_number> symlinks to /sys/devices/.../block/.../<deviceName>
+		symlink := filepath.Join(sysMountPoint, "dev", "block", deviceNumber)
+
+		discoveryLogger.Debug("evaluating sysfs symlink",
+			sglog.String("operation", "discovering device path"),
+			sglog.String("symlink", symlink),
+		)
+
+		devicePath, err := filepath.EvalSymlinks(symlink)
+		if err != nil {
+			discoveryLogger.Debug("failed to evaluate sysfs symlink",
+				sglog.String("operation", "discovering device path"),
+				sglog.Error(err),
+			)
+
+			continue
+		}
+
+		discoveryLogger.Debug("discovered device path",
+			sglog.String("operation", "discovering device path"),
+			sglog.String("devicePath", devicePath),
+		)
+
+		// Check to see if devicePath points to a disk partition. If so, we need to find the parent
+		// device.
+
+		for {
+			if devicePath == "" || devicePath == "/" || devicePath == "." {
+				// stop walking up the folder hierarchy once we have an empty path (or a terminal as defined by filepath.Dir)
+				break
```

Comment:
if we hit this case we should give up. It looks like you still do os.Stat

Reply:
Resolved in a different way, but the plan was to rely on the next os.Stat failing for us (example `os.Stat("/partition")`)